include/seqan3/alignment/matrix/alignment_trace_matrix.hpp

```diff
@@ -98,75 +99,148 @@ struct alignment_coordinate
+{
+    using signed_size_t = std::make_signed_t<size_t>;
+
+    constexpr auto N = trace_directions::none;
+    constexpr auto D = trace_directions::diagonal;
+    constexpr auto L = trace_directions::left;
+    constexpr auto U = trace_directions::up;
+    signed_size_t row = end_coordinate.seq2_pos + 1;
+    signed_size_t col = end_coordinate.seq1_pos + 1;
+
+    assert(row < matrix.rows());
+    assert(col < matrix.cols());
+
+    while(true)
+    {
+        trace_directions dir = matrix.at(row, col);
+        if ((dir & L) == L)
+        {
+            col = std::max<signed_size_t>(col - 1, 0);
+        }
+        else if ((dir & U) == U)
+        {
+            row = std::max<signed_size_t>(row - 1, 0);
+        }
+        else if ((dir & D) == D)
+        {
+            row = std::max<signed_size_t>(row - 1, 0);
+            col = std::max<signed_size_t>(col - 1, 0);
+        }
+        else
+        {
+            if (row == 0 || col == 0)
+                break;
+            throw std::logic_error{"Trace not found"};
+        }
+    }
+    return
+    {
+        std::max<signed_size_t>(col - 1, 0),
+        std::max<signed_size_t>(row - 1, 0)
+    };
+}
+
+/*!\brief Compute the trace from a trace matrix.
+ * \ingroup alignment_matrix
+ * \tparam    database_t                 The type of the database sequence.
+ * \tparam    query_t                    The type of the query sequence.
+ * \tparam    trace_matrix_t             The type of the trace matrix.
+ * \cond DEV
+ * \tparam    gapped_database_alphabet_t The alphabet type of the gapped database sequence.
+ * \tparam    gapped_query_alphabet_t    The alphabet type of the gapped query sequence.
+ * \endcond
+ * \param[in] database                   The database sequence.
+ * \param[in] query                      The query sequence.
+ * \param[in] matrix                     The trace matrix.
+ * \param[in] end_coordinate             Where the trace in the matrix ends.
+ */
 template <
     typename database_t,
     typename query_t,
     typename trace_matrix_t,
-    typename gapped_database_t = gapped<std::decay_t<decltype(std::declval<database_t>()[0])>>,
-    typename gapped_query_t = gapped<std::decay_t<decltype(std::declval<query_t>()[0])>>>
+    typename gapped_database_alphabet_t = gapped<std::decay_t<decltype(std::declval<database_t>()[0])>>,
+    typename gapped_query_alphabet_t = gapped<std::decay_t<decltype(std::declval<query_t>()[0])>>>
 //!\cond
     requires matrix_concept<remove_cvref_t<trace_matrix_t>> &&
              std::Same<typename remove_cvref_t<trace_matrix_t>::entry_type, trace_directions>
 //!\endcond
-inline std::pair<std::list<gapped_database_t>, std::list<gapped_query_t>>
-alignment_trace(database_t && database, query_t && query, trace_matrix_t && matrix, alignment_coordinate const end_coordinate)
+inline std::pair<std::vector<gapped_database_alphabet_t>, std::vector<gapped_query_alphabet_t>>
```

Comment:
return a tuple?

Reply:
and this must also be on the same line, then???

Reply:
tuple is not possible right now, because of #471. @smehringer has to fix the `aligned_sequence -> cigar` conversion function.